doc: external components: Introduce a section about external tooling

carlescufi opened the pull request:
Since tools that are not software that is compiled into the final images may have very different requirements from regular external components, introduce a new section outlining the process and rules to introduce new tools into the project.

carlescufi committed:
doc: external components: Introduce a section about external tooling

Since tools that are not software that is compiled into the final images
may have very different requirements from regular external components,
introduce a new section outlining the process and rules to introduce new
tools into the project.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

mbolivar-nordic commented on doc/contribute/external.rst:
```diff
@@ -164,3 +170,32 @@ The flowchart below shows an overview of the process:
+
+Just like with regular external components, tooling that is imported from
+another project can be integrated either in the main tree or as a :ref:`west
+project <west-workspace>` Note that in this case the corresponding west project
```
Missing period

mbolivar-nordic commented on doc/contribute/external.rst:
```diff
@@ -164,3 +170,32 @@ The flowchart below shows an overview of the process:
+another project can be integrated either in the main tree or as a :ref:`west
+project <west-workspace>` Note that in this case the corresponding west project
+will not be a :ref:`module <modules>`, because tooling does not make use of the
+Zephyr build system nor needs to be processed by it.
```
and does not need

mbolivar-nordic reviewed (commented):
Found a couple of typos.

nashif commented on doc/contribute/external.rst:
```diff
@@ -164,3 +170,32 @@ The flowchart below shows an overview of the process:
+manifest. More info on optional projects can be found in :ref:`this section
+<west-manifest-groups>`.
+
+The TSC must approve every Pull Request that introduces a new external tool.
```
I would add a note about importing manifests. Changes to the imported manifests, i.e. adding new projects or repos must be reviewed by the TSC as well. We should not have manifests go beyond what was initially approved.

carlescufi committed:
doc: modules: Clarify the difference between modules and projects

Since this has been a source of confusion for a while now, add a small
section to clarify the difference between modules and west projects.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: external components: Add a section about the main manifest

Add a generic section that covers changes to the main manifest,
requiring those to have TSC approval.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

galak commented on doc/contribute/external.rst:
```diff
@@ -164,3 +170,47 @@ The flowchart below shows an overview of the process:
+manifest. More info on optional projects can be found in :ref:`this section
+<west-manifest-groups>`.
+
+The TSC must approve every Pull Request that introduces a new external tool.
```
I think we need some guidelines on what will or will not be approved by the TSC for a tool here.

ifyall replied:
I agree. The only guidance we have right now is in https://github.com/zephyrproject-rtos/zephyr/pull/56191/commits/0011ff11022e0f9a8caac986efc396a45be87455, and as I commented there, there isn't any guidance on why a tool or its license may be rejected, even if there are not license requirements.

ifyall commented on doc/contribute/external.rst:
```diff
@@ -164,3 +170,33 @@ The flowchart below shows an overview of the process:
+of the code compiled and linked into the final image.
+
+There are no license requirements for tooling components, but licenses that
+have not been approved by the `Open Source Initiative (OSI)`_ might be rejected.
```
This feels a little strange. If there are no license requirements, why would a tool component be rejected due to its license?

galak reviewed (changes requested):
Need guidelines about what is acceptable or not for tooling.

aescolar reviewed (dismissed):
This looks good to me.
The first commit just clarifies that projects which are not modules exist.
The other 2 commits, apart from clarifying what is expected overall, clearly leave the decision to the TSC. And given that there could be many considerations we can't foresee today, leaving it open and for the TSC to judge seems reasonable to me.

nashif commented on doc/develop/modules.rst:
```diff
@@ -38,6 +38,32 @@ references to optional :ref:`binary blobs <bin-blobs>`.
+:ref:`West projects <west-manifests-projects>` are entries in the `projects:`
+section in the :file:`west.yml` manifest file.
+West projects are often also modules, but not always. There are west projects
+that do not require integration with the build system (eg. tools) and thus do
```
this is misleading, we might actually have a tool in a project that integrates with the build system, what you probably mean here '... does not end up as part of the built image", no? or would this be a module? :)

carlescufi commented:
> End users seeking any certification level may need to know the qualification level achieved by each tool they use. Within the Zephyr ecosystem this should be consistently reported. The guidelines on tool integration should cover how this is done.
> 
> Similarly, each tool should have some preliminary documentation within the Zephyr documentation structure. This documentation should exist within the repository of each tool to avoid extraneous content when the tool is not included in the workspace.

I really do not have enough knowledge to document or discuss this to be honest. Would you like to send a PR once this one is merged to add the requirements you mention?

galak commented on doc/contribute/external.rst:
```diff
@@ -164,3 +170,49 @@ The flowchart below shows an overview of the process:
+of the code compiled and linked into the final image.
+
+Tooling components must be released under a license approved by the
+`Open Source Initiative (OSI)`_.
```
Do we want to make any comment here about any exceptions should be raised to the TSC for initial discussion.  This makes it sound like if its not OSI its not ok, which isn't true (as we clearly support commercial tooling at various levels).

galak replied:
So we need to either clarify what this section applies to or explain how a commercial compiler applies.   If I was new to the project I would think that this section would say that the project isn't interested or willing to take support for that commercial compiler - since its not OSI licensed approved.

galak commented on doc/contribute/external.rst:
```diff
@@ -164,3 +170,49 @@ The flowchart below shows an overview of the process:
+`Open Source Initiative (OSI)`_.
+
+Just like with regular external components, tooling that is imported from
+another project can be integrated either in the main tree or as a :ref:`west
```
I'm not sure if this is clear w/regards to tools like a compiler for example, that isn't part of the main tree and isn't reference via west.

carlescufi replied:
But then for those tools this page is not applicable at all, because it’s not part of the “default distribution”. The docker image is not considered part of the zephyr distro

galak replied:
where do we define default distribution?

galak commented on doc/contribute/external.rst:
```diff
@@ -164,3 +170,49 @@ The flowchart below shows an overview of the process:
+
+The TSC must approve every Pull Request that introduces a new external tooling
+component. This will be done on a case-by-case, individual analysis of the
+proposed addition by the TSC representatives.
```
I still think we should describe more about what kinda of tooling would be found to be acceptable and what might not.

galak replied:
I think its about giving an example.  That something like a script utilizing pyelf is ok since its OSI approved, etc...

galak reviewed (changes requested):
I think there are still some areas that need clarification.

nashif commented:
approved by the TSC.
